qutip/core/cy/qobjevo.pyx

```diff
@@ -0,0 +1,1114 @@
+                self.funcQEvo = (<QobjEvo> Q_object).funcQEvo.copy()
+            else:
+                self.args = (<QobjEvo> Q_object).args
+                self.elements = (<QobjEvo> Q_object).elements
+                self.funcQEvo = (<QobjEvo> Q_object).funcQEvo
+            if args:
+                self.arguments(args)
+            return
+
+        self.elements = []
+        self.funcQEvo = set()
+        self.func_qobj = {}
+        self.dims = None
+        self.shape = (0, 0)
+        self.args = args.copy()
+        self._issuper = -1
+        self._shift_dt = 0
+
+        use_step_func = args.get("_step_func_coeff", 0)
+        cte = None
+
+        if (
+            isinstance(Q_object, list)
+            and len(Q_object) == 2
+            and isinstance(Q_object[0], Qobj)
+            and not isinstance(Q_object[1], (Qobj, list))
+        ):
+            # The format is [Qobj, f/str]
+            Q_object = [Q_object]
+
+        if isinstance(Q_object, Qobj):
+            self.elements = [_CteElement(Q_object)]
+            self.dims = Q_object.dims
+            self.shape = Q_object.shape
+
+        elif isinstance(Q_object, list):
+            for op in Q_object:
+                if isinstance(op, Qobj):
+                    if cte is None:
+                        cte = op
+                    else:
+                        cte += op
+                    _dims = op.dims
+                    _shape = op.shape
+                else:
+                    self.elements.append(_EvoElement(
+                        op[0],
+                        coefficient(op[1], tlist=tlist, args=args,
+                                    _stepInterpolation=use_step_func,
+                                    compile_opt=CompilationOptions())
+                    ))
+                    _dims = op[0].dims
+                    _shape = op[0].shape
+
+                if self.dims is None:
+                    self.dims = _dims
+                    self.shape = _shape
+
+                else:
+                    if self.dims != _dims:
+                        raise ValueError("incompatible dimensions " +
+                                         str(self.dims) + ", " + str(_dims))
+
+            if cte is not None:
+                self.elements = [_CteElement(cte)] + self.elements
+
+        elif callable(Q_object):
+            qobj = Q_object(0, args)
+            if not isinstance(qobj, Qobj):
+                raise ValueError("Function based time-dependent system must "
+                                 "have the signature "
+                                 "`f(t: double, args: dict) -> Qobj`")
+            self.dims = qobj.dims
+            self.shape = qobj.shape
+            self.funcQEvo = set([Q_object])
+            self.elements.append(_FuncElement(Q_object, qobj, args))
+
+        else:
+            raise TypeError("Format not understood")
+
+    def __call__(self, t, args={}):
```

Comment:
I am thinking of removing the option and have `__call__` accept only `t`. Since each coefficient has a copy of args, `QobjEvo` keep a copy only for this functionality. It usage should be rare enough that asking the user to update args with arguments seems reasonable.

Reply:
The way I think about `args` at different scoping levels, I like having it available to pass to `__call__` here, but I'm not too beaten up if you think it would be much better to remove it.  It'd probably be good to have benchmarks to know if it's actually worthwhile doing, though, since we're changing the API.

Regardless, as long as it's in here, a default value of `{}` is unsafe - it's always unsafe in Python to have a default parameter be mutable.

Reply:
The issue I see is if you have coefficient with different value of the same key in args, they would lose there unique values  with this...  `Coeffcient.__call__` also accept an input args but it overwrite the dict and does not come back to the previous state. I will keep the function by making a copy, this would ensure that the original args are untouched, but it may be slower.

Reply:
So my main point about the kwarg default needs to be `args=None` I think is fixed, so I'm fine on this thread.

For the other bits: I think my actual concern is more about how `args` is handled _everywhere_.  Currently, you can pass an `args` dictionary at various points, which feels like you're defining `args` within that scope.  However, what really happens is that an underlying `Coefficient` object gets mutated (no matter how many references may exist to it), and the values are never restored afterwards.  Now that I think about it, this is already causing behaviour that's incorrect:
```python
>>> my_td = qutip.coefficient('w + 0*t', args={'w': 1})
>>> my_obj = qutip.QobjEvo([[qutip.qeye(2), my_td]])
>>> my_obj(0) == qutip.qeye(2)
True
>>> # now let's do some evolution
>>> H = [[qutip.sigmax(), my_td]]
>>> res = qutip.mesolve(H, qutip.basis(2, 0).proj(), [0, 1], args={'w': 2})
>>> # now let's look at my_obj again
>>> my_obj(0) == qutip.qeye(2)
False
>>> my_obj(0) == 2 * qutip.qeye(2)
True
```

I think we really very much need to avoid implicit mutation like this.  Perhaps we need to discuss/design this a bit more, because this unexpected mutation absolutely will cause people problems, since we're really billing `Coefficient` as compile once, use many times.

To hook it back to the point you bring up, though: having `__call__` mutate any accessed `Coefficient` seems quite clearly wrong (hence you putting in the `copy`), but to me it's equally wrong to have `mesolve` (and anything else) do the same thing as well.  This is beyond the scope of this PR for sure, but it's something I think will really bite us and users if we leave it as-is.

Reply:
I wouldn't remove `args` from the signature of any of these points, I'd make it so that _nothing_ except a user's direct call to `Coefficient.arguments()` can mutate an object that they have a reference to.  This becomes a little more complicated when dealing with `QobjEvo` - a user's call to `QobjEvo.arguments()` seems ok to mutate held `Coefficient` objects, but passing the `QobjEvo` to `mesolve` shouldn't risk any mutation at all.

Reply:
Can we resolve this thread now that `Coefficients` are immutable?